ql/src/java/org/apache/hadoop/hive/ql/scheduled/ScheduledQueryExecutionService.java

```diff
@@ -183,33 +232,34 @@ private String getErrorStringForException(Throwable t) {
 
     @Override
     public void run() {
-      while (true) {
+      while (!context.executor.isShutdown()) {
         try {
           Thread.sleep(context.getProgressReporterSleepTime());
         } catch (InterruptedException e) {
           LOG.warn("interrupt discarded");
         }
         try {
-          worker.reportQueryProgress();
+          for (ScheduledQueryExecutor worker : runningWorkers) {
```

Comment:
Can we hit a concurrent modification exception if a worker finishes and is deleted, while we are iterating here?

Reply:
yes; it was a simple LinkedList - and for that there was a slight chance...I've switched to ConcurrentLinkedQueue to avoid that.